ctypesgen/printer_python/preamble.py

```diff
@@ -1,7 +1,16 @@
+from typing import TypeVar, Type, get_type_hints
 import ctypes
 import sys
 from ctypes import *  # noqa: F401, F403
 
+T = TypeVar("T")
+
+def struct_decorator(cls: Type[T]) -> Type[T]:
+    class Internal(cls.__base__):
```

Comment:
Why are we inheriting from `__base__` here, and not from `cls` directly?
Wouldn't that lose the specific `struct_...` type, allowing to pass mismatched structs to functions?

Reply:
Ah, I think it doesn't allow that, since this is a class factory, so each returned `Internal` is actually an own new type. However, if we construct a struct, wouldn't `type(struct_obj)` then show `Internal`, instead of `struct_...`?

Anyway, why this wrapper at all? Why don't we just define these directly in `print_struct()` ??

Reply:
Ah, doing some testing I figured out `__slots__` prevent assignment of bad fields if, and only if defined on the direct inheritance from `ctypes.Structure`.

Anyway, although I now understand the background, the `__base__` inheritance is still likely to cause confusion: it loses anything defined in the class body below, unless taken over explicitly here.

Reply:
As you guessed correctly, this wrapper exists solely as a method of adding slots without repetition. There is probably a better way, but I'm not aware.
I agree that stripping anything but the annotations would be confusing in a normal setting. However, this is part of generated code, where we define structs as part of an API for which (afaik) nothing else matters. If you fear misuse by lib users, we could `del` the decorator [1]. Otherwise we should probably not use slots by default thus rendering the wrapper obsolete.

[1] In general I would welcome it if ctypesgen only exposed stuff from the actual lib.

Reply:
Erm, but why do we need the wrapper at all? Can't you just access `__annotations__` (without dotted prefix) directly in the class body, after all annotations were defined?
```pycon
>>> class Test:
...     a: int
...     _test = __annotations__
... 
>>> Test._test
{'a': <class 'int'>}
>>> 
```

Reply:
Yes, losing members that aren't taken over explicitly does not affect the downstream result, but it'd just be incredibly confusing for a future ctypesgen developer.
Apart from that, the problem of losing the name remains, which *does* affect downstream users' debugging experience.

Reply:
> Erm, but why do we need the wrapper at all? Can't you just access `__annotations__` (without dotted prefix) directly in the class body?

For avoiding repetition in `__slots__` (see comment above).

> Yes, `__base__` inheritance losing members does not affect the downstream result as long as they're correctly taken over in ctypesgen, but it'd just be incredibly confusing for a future ctypesgen developer. (Also the takeover may caus duplication, too, though of a different sort.)

This can be fixed by adding a comment

> Apart from that, the problem of losing the name remains, which _does_ affect downstream users' debugging experience.

I just added the fix mentioned in the edit of my comment above.

Reply:
Nice we got there in the end. I think we've resolved quite some mutual confusion in this thread 😅